File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stk = []
        for ch in s:
            if ch in ['(', '[', '{']:
               stk.append(ch)
            elif stk and (stk[-1], ch) in [tuple('()'), tuple('[]'), tuple('{}')]:
                stk.pop()
            else:
                return False
        return not stk

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        var stk = new Stack<Character>();
        
        for(var ch : s.toCharArray()) {
            if (List.of('(', '[', '{').contains(ch)) {
                stk.add(ch);
                continue;
            }
            if (stk.isEmpty()) {
                return false;
            }
            if (stk.peek() == '(' && ch == ')') {
                stk.pop();
                continue;
            }
            if (stk.peek() == '[' && ch == ']') {
                stk.pop();
                continue;
            }
            if (stk.peek() == '{' && ch == '}') {
                stk.pop();
                continue;
            }
            return false;
        }
        return stk.empty();
    }
}
